cpp/src/strings/strings_column_factories.cu

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <bitmask/valid_if.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/functions.h>
+#include <cudf/null_mask.hpp>
+#include <cudf/strings/strings_column_factories.hpp>
+#include <utilities/error_utils.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/transform_reduce.h>
+#include <thrust/transform_scan.h>
+#include <thrust/for_each.h>
+
+
+namespace cudf {
+
+// Create a strings-type column from array of pointer/size pairs
+std::unique_ptr<column> make_strings_column(
+    const rmm::device_vector<thrust::pair<const char*,size_t>>& strings,
+    cudaStream_t stream,
+    rmm::mr::device_memory_resource* mr)
+{
+    size_type count = strings.size();
+    // maybe a separate factory for creating null strings-column
+    CUDF_EXPECTS(count > 0, "must specify at least one pair");
+
+    auto execpol = rmm::exec_policy(stream);
+    auto d_strings = strings.data().get();
+
+    // check total size is not too large for cudf column
+    size_t bytes = thrust::transform_reduce( execpol->on(stream),
+        thrust::make_counting_iterator<size_t>(0),
+        thrust::make_counting_iterator<size_t>(count),
+        [d_strings] __device__ (size_t idx) {
+            auto item = d_strings[idx];
+            return item.first ? item.second : 0;
+        },
+        0, thrust::plus<size_t>());
+    CUDF_EXPECTS( bytes < std::numeric_limits<size_type>::max(), "total size of strings is too large for cudf column" );
+
+    // build offsets column
+    auto offsets_column = make_numeric_column( data_type{INT32}, count, mask_state::UNALLOCATED, stream, mr );
+    auto offsets_view = offsets_column->mutable_view();
+    thrust::transform_inclusive_scan( execpol->on(stream),
```

Comment:
I think computing the offsets should use an *exclusive* scan. The offsets array should be one longer than the number of strings in the column, so that the last element is the offset past the end of the last string. This would avoid conditional logic every time the offset is accessed. Currently, a lot of locations in the strings code do something like `offset = (index == 0) ? 0 : d_offsets[index-1]`. 

With an exclusive scan offset, that just becomes `offset = d_offsets[index]` and length is just `d_offsets[index+1] - d_offsets[index]`.  

In my experience storing offsets starting with 0 (exclusive scan) is common. Moreover, it would make strings columns consistent with arrow list layout.  
https://arrow.apache.org/docs/memory_layout.html (Search for "List Type")

Reply:
Storing the 0 offset seems a waste of space. Space that can allow storing one more string if we are around the size_type limit.
Regardless, there is no place to store the number of strings in the column. The parent column holds the nulls and the children but no data. The chars child column is the number of bytes for the chars. The offsets column is the number of offsets and therefore the number of strings. The parent column cannot have a non-zero size count along with a null data-pointer. This means the only place to store the accurate number of strings is the offsets column.
So for this reason (and not wanting to storing an eternal 0) the offsets match the number of strings.
If we could allow a column to have a size with no data, this may work but I think it may be hard to support.

Reply:
> there is no place to store the number of strings in the column.

The parent _must_ hold the size, if not, if you were to try and pass a `cudf::column` of type `STRING` into a `table` constructor, it would fail because the column would have zero size. 

> If we could allow a column to have a size with no data, this may work but I think it may be hard to support.

Agreed, if we just do this, then the parent can hold the number of elements. I don't think it should be that hard, I already have a special case for the `EMPTY` column type. In fact, if we decide on a pattern for compound column types where the parent never contains data, then it could be a simple check for `is_compound` to allow a non-zero size with null data.

Reply:
I thought a `non-zero size with no data` column could represent a column with all nulls. So it would have a size (number of elements) by not actually require any data. We should probably define what a null/empty column is and what a column with all nulls. A STRING column must also support a column with all empty strings (not-null). This also has no data but maybe just all offsets are 0.

Reply:
> I thought a non-zero size with no data column could represent a column with all nulls.

This is currently only true for the `EMPTY` type.